Autopilot docs

fedgiac opened the pull request:
# Description

Documents how the autopilot works.

# Changes

- Filled in autopilot page.
- Simplified architecture page (separate `/reveal` and `/settle` is too much information there imho).

## Related Issues

Part of https://github.com/cowprotocol/docs-v2/issues/41.

fedgiac committed:
Autopilot docs

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
 ---
 sidebar_position: 2
-draft: true
```
@mfw78 How can we make it show up above Driver and Solver Engine in the sidebar?

![image](https://github.com/cowprotocol/docs-v2/assets/1200333/918843f7-adb8-4aaa-a454-abafd9078ab5)

mfw78 replied:
The way to do this is either in the `category.json`, or in the front-matter (the information between the `---` at the top of the YAML). By setting `sidebar_position`, the docs can be ordered this way. I've gone through and pushed these changes to order the docs correctly on this branch.

fedgiac replied:
For reference, this was done in commit b3089c2dd35913b4f65ff9b16e999d658a662429.

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+Its role can be broadly summarized into these main purposes.
+
+1. Data collection, which is mainly:
+   - Liquidity gathering
```
The autopilot no longer gathers liquidity

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+broadcasts this auction to each of the solvers,
+and finally chooses which solution will be executed.
+
+Users don't interact with the autopilot directly: its only intended interface is with the solvers (through the driver).
```
Link to the driver (which should be the next logical page on the chain through the system)

```suggestion
Users don't interact with the autopilot directly: its only intended interface is with the solvers (through the [driver](driver)).
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+and finally chooses which solution will be executed.
+
+Users don't interact with the autopilot directly: its only intended interface is with the solvers (through the driver).
+Unlike solvers, there is a single autopilot running for each chain[^barn].
```
Not sure how this relates to solvers?
```suggestion
There is a single autopilot running for each chain[^barn].
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+   - Liquidity gathering
+   - Order information
+2. Cutting auctions
+3. Solver competition
```
```suggestion
3. Solver competition
  - Data availability consensus around the available solution candidates and their score
  - Notifying the winning solver to submit their solution
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+
+Its role can be broadly summarized into these main purposes.
+
+1. Data collection, which is mainly:
```
```suggestion
1. Cutting auctions:
  - Data availability consensus around which orders are valid for a given batch
  - The initial exchange rates for each traded tokens which are used to normalize surplus across different orders
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+Running at a regular interval, it keeps an up-to-date view on the state of the protocol,
+synthesizes this data into an _auction_,
+broadcasts this auction to each of the solvers,
+and finally chooses which solution will be executed.
```
are all these newlines intended?

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+    participant database
+    participant blockchain
+    participant autopilot
+    box external drivers
```
As far as the autopilot is concerned it's communicating with solvers (which usually consist of a driver and solver engine), but this is an implementation detail from the autopilot perspectives and hopefully solvers will soon build their own integrated solvers that serve both purposes.

```suggestion
    box external solvers
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
````diff
@@ -1,6 +1,150 @@
+
+```mermaid
+sequenceDiagram
+    participant orderbook
````
I'd remove the orderbook as a participant as it doesn't really play a direct role in the lifecycle of the auction (we can add in writing that the database is populated by the orderbook)

MartinquaXD replied:
I'd also rename `driver` to `solver` and drop one of them. Having 2 should be sufficient to understand the `1:n` relationship without bloating the diagram.

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+      blockchain->>-autopilot: Onchain cancellations, ETH flow...
+    and
+      autopilot->>+blockchain: Get liquidity
+      blockchain->>-autopilot: Uniswap, Balancer, ...
```
Liquidity is now fetched by the driver and only if configured as such. The only reason the autopilot still depends on on-chain liquidity is for native price estimates (which should mid-term also be done by the connected solvers)

```suggestion
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+
+## Data collection
+
+The autopilot needs to have a complete overview of the protocol and of many other third-party protocols in order to build an [auction](/cow-protocol/reference/core/auctions/schema) that is as complete as possible.
```
What other third party protocols does the autopilot need to have an overview of?

fedgiac replied:
This was intended to be about liquidity, but since it isn't relevant anymore I removed this direct references to external protocols.

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+- Uniswap (v2 and v3)
+- Balancer (Weighted, Weighted2Token, ComposableStable, ...)
+- SushiSwap
+- Swapr
```
This needs a re-write

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+
+The autopilot also checks that [ERC-1271](/cow-protocol/reference/core/signing-schemes#erc-1271) signatures are currently valid.
+
+This steps tries to guarantees that all available orders can be settled on CoW Protocol without further checks from the solvers.
```
I thought we are less strict now about the filtering at the autopilot level (ie. if a user has 1 WETH but two open orders selling 1WETH, we forward both orders and let the driver do the advanced filtering)

MartinquaXD replied:
We filter orders partially in the `autopilot`. If a user only has 0.5 ETH we filter out all orders that sell more than that because there is absolutely no way they can be solved. (for partially fillable orders we only require the user to have at least 1 wei)
But if that user has 10 0.4 ETH orders we forward all of them to the solvers and let them figure out which one to prioritize.
So we filter out some orders but that does not yet guarantee that a solver can pick and solve any number of the orders without checks.

fedgiac replied:
I removed this line.
It feels like there should be an explanation here on why we prune the orders and we don't let the solvers get all of them regardless of these checks. (For example, a solver might be able to see that, after settling an order, a user would have enough balance to settle another one.)
I'm not sure what to write, however.

MartinquaXD replied:
> (For example, a solver might be able to see that, after settling an order, a user would have enough balance to settle another one.)

This case doesn't really work, though. All orders happen atomically so there is no before and after within the same block.

> I'm not sure what to write, however.

I would write something like the `autopilot` only filters out orders that are guaranteed to be unsettleable. All orders where it's ambiguous/unclear get forwarded to the solver to let it decide instead.

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+
+## Solver competition
+
+Once an auction is ready, the autopilot sends a `/solve` request to each solver's driver.
```
For the protocol perspective we should not leak the differentiation between driver and solver engine (it's just a solver)

```suggestion
Once an auction is ready, the autopilot sends a `/solve` request to each solver.
```

fleupold commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+## Complexities
+
+A typical challenge in the autopilot is handling block reorgs.
+The autopilot must be able to revert as many actions as possible in case of a reorg; everything that can't be reverted must be accounted for in the stored data.
```
```suggestion
The autopilot must be able to revert as many actions as possible in case of a reorg; everything that can't be reverted must be accounted for in the stored data.

In practice this means that some information (e.g. competition data by transaction hash) is only available after a "reorg safe" threshold of blocks have been proposed.
```

fleupold reviewed (commented):
Looks good. I think there are some inaccuracies as to what the autopilot does in the new model (ie not fetch liquidity any more, do less filtering).

From the autopilot's point of view we shouldn't differentiate between driver and solver-engines (it's abstraction is the "solver", which in practice consists of driver and solver-engine).

fedgiac committed:
Remove direct references to the driver in favor of solvers

MartinquaXD commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+## Solver competition
+
+Once an auction is ready, the autopilot sends a `/solve` request to each solver's driver.
+Solvers have a short amount time (seconds) to come up with a [solution](/cow-protocol/core/auctions/the-problem#solution) and return its _score_ to the autopilot.
```
Should we give a broad explanation of what the `score` is?
We could change the exact details of the score so we should probably not mention a formular but maybe saying that it roughly describes the efficiency, quality, goodness (or sth like that) of a solution would be useful.

fedgiac replied:
I expanded a bit but I think it's important and independent enough that it deserves its own page. It would be nice to have a dedicated page explaining why the score is what it is and the research that went into it.

MartinquaXD commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+The autopilot selects the winner based on this score once the allotted time expires or all drivers have returned their batch proposal.
+
+Up to this point, the autopilot only knows the score and not the full solution that achieves that score.
+The autopilot then asks the winning solver to reveal its score (throug `/reveal`) and then to execute the corresponding settlement transaction (`/settle`).
```
```suggestion
The autopilot then asks the winning solver to reveal its score (through `/reveal`) and then to execute the corresponding settlement transaction (`/settle`).
```

MartinquaXD commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+
+Up to this point, the autopilot only knows the score and not the full solution that achieves that score.
+The autopilot then asks the winning solver to reveal its score (throug `/reveal`) and then to execute the corresponding settlement transaction (`/settle`).
+The solver is responsible for executing the transaction onchain (through the driver).
```
Making the differentiation of `driver` + `solver` is not needed from the perspective of the `autopilot`.

MartinquaXD commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,150 @@
+
+## Auction data storage
+
+The data returned by the solver is stored by the autopilot on the database.
```
```suggestion
The data returned by the solver is stored by the autopilot in the database.
```

fedgiac committed:
Phrasing comments from Felix

fedgiac committed:
Expand on how reorgs are handled

MartinquaXD reviewed (commented):
Over all quite nice. But I think we should no longer mention liquidity and don't make the differentiation of `solver` and `driver` from the `autopilot`'s perspective.

fedgiac committed:
Remove orderbook from diagram

fedgiac committed:
Remove mentions to third-party protocols

These were ment to be liquidity sources, but it isn't too relevant to the protocol anymore.

fedgiac committed:
Merge auction cutting and data collection

fedgiac committed:
Remove direct reference to liquidity, rather expanding on price estimators

fedgiac committed:
Expand on solver competition summary

fedgiac committed:
Remove comment on auction guarantees

fedgiac committed:
Expand on score

fedgiac committed:
Typos from Martin

mfw78 committed:
chore: arbitrate ordering

fedgiac committed:
Expand on what the autopilot ignores when creating an auction

mfw78 commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,140 @@
+The reference token is usually the chain's native token, since it's the token used to pay for the gas needed when executing a transaction. 
+Orders whose price can't be fetched are discarded and won't be included in an auction.
+
+Native token price fetching is handled by a integrated price estimator in the autopilot.
```
```suggestion
Native token price fetching is handled by an integrated price estimator in the autopilot.
```

mfw78 commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,140 @@
+
+Native token price fetching is handled by a integrated price estimator in the autopilot.
+The price is fetched from multiple sources and may change based on the current configurations.
+Prices are both queried from a list of selected existing solvers as well as retrieved internally to the autopilot (for example, by querying some external parties like Paraswap and 1inch, but also by reading onchain pool data as Uniswap).
```
```suggestion
Prices are both queried from a list of selected existing solvers as well as retrieved internally by the autopilot (for example, by querying some external parties like Paraswap and 1inch, but also by reading onchain pool data as Uniswap).
```

mfw78 commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,140 @@
+The price is fetched from multiple sources and may change based on the current configurations.
+Prices are both queried from a list of selected existing solvers as well as retrieved internally to the autopilot (for example, by querying some external parties like Paraswap and 1inch, but also by reading onchain pool data as Uniswap).
+
+Orders that can't be settled are filtered out: these are expired orders, those with not enough balance, with missing approvals, or that use tokens that aren't supported by the protocol.
```
```suggestion
Orders that can't be settled are filtered out, such as:
* expired orders
* insufficient balances
* missing approvals; or 
* that use tokens that aren't supported by the protocol
```

fedgiac replied:
I rephrased a bit to make it sound nicer in a list form, but I've otherwise applied this suggestion.

mfw78 commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,140 @@
+
+Once an auction is ready, the autopilot sends a `/solve` request to each solver.
+Solvers have a short amount time (seconds) to come up with a [solution](/cow-protocol/reference/core/auctions/the-problem#solution) and return its _score_ to the autopilot.
+The score is the quality of the solution, which accounts for surplus and fee paid, and represents what CoW Protocol intends to optimize with the solver competition; it's described in details in the [description of CoW Protocol's optimization problem](/cow-protocol/reference/core/auctions/the-problem).
```
I'd suggest compressing this further such that all details are farmed out to `the-problem` / technical reference and reduce duplication as much as reasonably practicable.

fedgiac replied:
Done. I'd like to keep the word "quality" in there because the linked page uses that instead of "score". Let me know if I should simplify further.

mfw78 commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -1,6 +1,140 @@
+
+Up to this point, the autopilot only knows the score and not the full solution that achieves that score.
+The autopilot then asks the winning solver to reveal its score (through `/reveal`) and then to execute the corresponding settlement transaction (`/settle`).
+The solver is responsible for executing the transaction onchain (through the [driver](driver)).
```
```suggestion
The solver is responsible for executing the transaction onchain (through the [driver](driver) if using the reference implementation).
```

mfw78 reviewed (approved):
Minor nits

fedgiac committed:
Typos from mfw78

fedgiac committed:
Explicit list of condition for filtered orders

fedgiac committed:
Specify that driver is an implementation detail

fedgiac committed:
Simplify score description

mfw78 commented on docs/cow-protocol/tutorials/arbitrate/autopilot.md:
```diff
@@ -107,13 +111,13 @@ Still, this doesn't mean that all orders that appear in the auction can be settl
-Solvers have a short amount time (seconds) to come up with a [solution](/cow-protocol/reference/core/auctions/the-problem#solution) and return its _score_ to the autopilot.
-The score is the quality of the solution, which accounts for surplus and fee paid, and represents what CoW Protocol intends to optimize with the solver competition; it's described in details in the [description of CoW Protocol's optimization problem](/cow-protocol/reference/core/auctions/the-problem).
-The autopilot selects the winner based on this score once the allotted time expires or all solvers have returned their batch proposal.
+Solvers have a short amount time (seconds) to come up with a [solution](/cow-protocol/reference/core/auctions/the-problem#solution) and return its _score_ to the autopilot, which represents the quality of a solution.
```
```suggestion
Solvers have a short amount of time (seconds) to come up with a [solution](/cow-protocol/reference/core/auctions/the-problem#solution) and return its _score_ to the autopilot, which represents the quality of a solution.
```

mfw78 reviewed (approved):
minor nit

fedgiac committed:
More typos found by mfw78
